File: backend/audiences.py
"""Audience configuration, shared with the UI.

One file describes all four audiences -- their labels, suggestions, source
collections and prompt text -- and both halves of the app read it:
`UI/src/config/audiences.ts` imports it, and this module loads the same bytes.
Duplicating it in two languages is a correctness bug waiting to happen, because
the prompt half and the UI half would drift silently and nothing would fail.

**Where it lives.** The JSON sits under `UI/src/config/` rather than a
repo-root `shared/` directory. Vite's root is `UI/`, and `UI/vite.config.js`
sets no `server.fs.allow` and no alias, so a file outside that root is an
unverified build risk for a gain of tidiness only. The backend reaching into
`UI/` is the uglier half of that trade and is made explicit here rather than
hidden. `AUDIENCES_CONFIG_PATH` overrides the location for a deployment that
ships the backend without the UI tree.

**Fails loudly.** The values here become prompt text and search scoping, so a
missing or malformed file is not something to paper over with a default.

The four ids are exactly the values the database check constraints allow on
`profiles.active_audience`, `profile_audience_details.audience` and
`conversations.audience`. `backend/tests/test_audiences.py` asserts that
against the migration, so adding a fifth fails the suite rather than failing in
the browser as "save failed".
"""
import json
import os
from functools import lru_cache
from pathlib import Path
from typing import Any

# Kept in sync with 20260917000100_audience_model.sql's three check
# constraints. The test compares against the migration text, not against this
# tuple, so this is a convenience rather than the authority.
AUDIENCE_IDS = ("student", "alumni", "guest", "faculty")

# The six values of the affiliation_kind enum. A different concept: what a
# person claims to be, not which experience they get.
AFFILIATION_KINDS = (
    "student",
    "alumni",
    "faculty",
    "staff",
    "applicant",
    "community",
)

_DEFAULT_PATH = Path(__file__).resolve().parent.parent / "UI" / "src" / "config" / "audiences.json"


def config_path() -> Path:
    """Where the shared config is read from. Env wins, for deployments."""
    override = os.getenv("AUDIENCES_CONFIG_PATH", "").strip()
    return Path(override) if override else _DEFAULT_PATH


@lru_cache(maxsize=1)
def _config() -> dict[str, Any]:
    """Parse the shared JSON once.

    Cached rather than read at import so the path can be overridden in a test
    with patch.dict(os.environ, ...) -- the idiom the rest of this suite uses --
    by calling `reload()` first. Import-time reads are what make
    runtime.py's constants untestable.
    """
    path = config_path()
    try:
        raw = path.read_text(encoding="utf-8")
    except OSError as exc:
        raise RuntimeError(
            f"audience configuration is missing at {path}. It is shared with "
            "UI/src/config/audiences.ts; set AUDIENCES_CONFIG_PATH if the "
            "backend is deployed without the UI tree."
        ) from exc

    try:
        parsed = json.loads(raw)
    except ValueError as exc:
        raise RuntimeError(f"audience configuration at {path} is not valid JSON") from exc

    audiences = parsed.get("audiences")
    if not isinstance(audiences, dict) or not audiences:
        raise RuntimeError(f"audience configuration at {path} has no 'audiences' object")

    missing = [a for a in AUDIENCE_IDS if a not in audiences]
    if missing:
        raise RuntimeError(
            f"audience configuration at {path} is missing: {', '.join(missing)}"
        )

    return parsed


def reload() -> None:
    """Drop the cached parse. For tests and for AUDIENCES_CONFIG_PATH changes."""
    _config.cache_clear()


def default_audience() -> str:
    return _config().get("default", "student")


def all_audiences() -> dict[str, dict]:
    return _config()["audiences"]


def is_audience(value: str | None) -> bool:
    """True when the value is one the database will accept."""
    return bool(value) and value in all_audiences()


def get(audience: str | None) -> dict:
    """One audience's config, falling back to the default.

    Never raises on an unknown id. `ChatRequest.audience` is client-supplied
    and the server must not trust it for anything authorization-shaped -- it
    selects prompt text and source collections only -- so an unrecognised value
    degrades to the default rather than erroring.
    """
    audiences = all_audiences()
    if audience and audience in audiences:
        return audiences[audience]
    return audiences[default_audience()]


def prompt_context(audience: str | None) -> str:
    return get(audience).get("promptContext", "")


def offices(audience: str | None) -> list[str]:
    return list(get(audience).get("offices", []))


def source_collections(audience: str | None) -> list[str]:
    """Domains this audience's retrieval should prefer and scope to."""
    collections = get(audience).get("sourceCollections") or []
    return [c for c in collections if isinstance(c, str) and c]


def site_filter(audience: str | None) -> str:
    """The `site:` fragment for a search query.

    One domain yields `site:sjsu.edu`; several yield
    `(site:a OR site:b)`. DuckDuckGo's `site:` already matches subdomains, so
    only distinct registrable domains belong in sourceCollections.
    """
    domains = source_collections(audience)
    if not domains:
        return ""
    if len(domains) == 1:
        return f"site:{domains[0]}"
    return "(" + " OR ".join(f"site:{d}" for d in domains) + ")"


def affiliation_to_audience() -> dict[str, str]:
    return dict(_config().get("affiliationToAudience", {}))

File: UI/src/config/audiences.ts
/**
 * Typed access to the audience configuration.
 *
 * `audiences.json` is the single source of truth and is also read by
 * `backend/audiences.py`, so the prompt half and the UI half cannot drift.
 * `backend/tests/test_audiences.py` asserts the two sides agree and that the
 * four ids match the database check constraints.
 *
 * This file is `.ts` rather than `.js` on purpose: `tsconfig.app.json` includes
 * only `src/**​/*.ts(x)`, so a `.js` config module would be type-checked by
 * nothing -- which is already true of every `.jsx` component and is not worth
 * extending.
 */
import raw from './audiences.json';

export type AudienceId = 'student' | 'alumni' | 'guest' | 'faculty';

/** The six values of the `affiliation_kind` enum. A claim, not an experience. */
export type AffiliationKind =
  | 'student'
  | 'alumni'
  | 'faculty'
  | 'staff'
  | 'applicant'
  | 'community';

export interface Suggestion {
  title: string;
  subtitle: string;
  prompt: string;
}

export interface ProfileField {
  name: string;
  label: string;
  type: 'text' | 'select' | 'month' | 'number' | 'tel' | 'email';
  icon?: string;
  placeholder?: string;
  options?: string[];
  /** Spans both columns of the two-column grid. */
  span?: number;
  maxLength?: number;
  min?: number;
  max?: number;
  step?: number;
  /**
   * The `profiles` column this field writes to. Present only for fields that
   * predate the audience model; everything else goes to
   * `profile_audience_details.details` as jsonb.
   *
   * A column named here MUST be in the UPDATE allowlist granted by
   * supabase/migrations/20260917000200_profiles_privilege_guard.sql, because
   * PostgREST rejects the whole statement if any one named column is ungranted.
   */
  column?: string;
}

export interface Audience {
  id: AudienceId;
  label: string;
  blurb: string;
  welcomeHeadline: string;
  /** Which `affiliation_kind` values fold into this audience. */
  affiliations: AffiliationKind[];
  /** What onboarding declares for this audience; null writes no row. */
  declaresAffiliation: AffiliationKind | null;
  sourceCollections: string[];
  offices: string[];
  promptContext: string;
  suggestions: Suggestion[];
  profileFields: ProfileField[];
}

interface AudienceConfig {
  default: AudienceId;
  audiences: Record<AudienceId, Audience>;
  affiliationToAudience: Record<AffiliationKind, AudienceId>;
}

const config = raw as unknown as AudienceConfig;

export const AUDIENCES: Record<AudienceId, Audience> = config.audiences;
export const DEFAULT_AUDIENCE: AudienceId = config.default;
export const AUDIENCE_IDS: AudienceId[] = Object.keys(AUDIENCES) as AudienceId[];
export const AFFILIATION_TO_AUDIENCE: Record<AffiliationKind, AudienceId> =
  config.affiliationToAudience;

/**
 * Resolve an audience id from whatever the caller has, falling back rather
 * than throwing. `profiles.active_audience` is `not null default 'student'`,
 * but a guest has no profile row at all and an older row may predate a value.
 */
export function resolveAudience(id: string | null | undefined): Audience {
  if (id && Object.prototype.hasOwnProperty.call(AUDIENCES, id)) {
    return AUDIENCES[id as AudienceId];
  }
  return AUDIENCES[DEFAULT_AUDIENCE];
}

/** True when `id` is one of the four the database will accept. */
export function isAudienceId(id: string | null | undefined): id is AudienceId {
  return !!id && Object.prototype.hasOwnProperty.call(AUDIENCES, id);
}

/** The audience a declared affiliation belongs to. */
export function audienceForAffiliation(affiliation: AffiliationKind): AudienceId {
  return AFFILIATION_TO_AUDIENCE[affiliation] ?? DEFAULT_AUDIENCE;
}

/**
 * The audiences a person may pick during onboarding, in display order.
 * Guest is included: someone may sign up and still only want public
 * information.
 */
export const ONBOARDING_ORDER: AudienceId[] = ['student', 'alumni', 'faculty', 'guest'];
